File: Data Structures & Algorithms/same-binary-tree/submission-8.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        pq = deque([p])
        qq = deque([q])
        while pq and qq:
            for _ in range(len(pq)):
                p_node = pq.popleft()
                q_node = qq.popleft()
                if p_node is None and q_node is None:
                    continue
                if p_node is None or q_node is None or p_node.val != q_node.val:
                    return False
                pq.append(p_node.left)
                qq.append(q_node.left)
                pq.append(p_node.right)
                qq.append(q_node.right)
        return True

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-8.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn lowest_common_ancestor(
        root: Option<Rc<RefCell<TreeNode>>>,
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        let mut curr = root;
        let p_val = p.as_ref()?.borrow().val;
        let q_val = q.as_ref()?.borrow().val;
        while let Some(node) = curr {
            let node_val = node.borrow().val;
            if p_val < node_val && q_val < node_val {
                curr = node.borrow().left.clone();
            } else if p_val > node_val && q_val > node_val {
                curr = node.borrow().right.clone();
            } else {
                return Some(node);
            }
        }
        None
    }
}
